### PR TITLE
[incubator/vault] Initial commit of vault chart

### PR DESCRIPTION
**Pull request by jpds**

### CODE REVIEW

File: incubator/vault/values.yaml
Comment (mumoshu) [inline incubator/vault/values.yaml]: @jpds Hi, thanks for the great PR!
It could sound a nit but can we make this part uncommented and modified like:

```yaml
# A YAML representation of a final vault config.json file.
# See https://www.vaultproject.io/docs/configuration/index.html for more information
vault:
  listener:
    tcp:
      address: '0.0.0.0:8200'
      tls_disable: 1
  storage:
    consul:
      address: ""
      path: ""
```

so that for a basic use-case we can just override `consul.address` via the `--set` flag like `helm install stable/vault --set vault.storage.consul.address="myconsul-svc-name"`?

**Comment by jpds** [comment]
Hello @mumoshu, thanks for the review - I've gone ahead and added those suggestions in and an alternative config in README too.

**Comment by kachkaev** [comment]
This is awesome!

@jpds what do you think of adding [vault-ui](https://github.com/djenriquez/vault-ui) as an optional component that would be available through ingress? Setting `ui.enabled` will turn it on and then adding things like `ui.ingress.xxx` will tweak the details. Or should there be a separate chart called `vault-ui` that depends on this chart (like wordpress depends on mariadb)?

**Comment by jpds** [comment]
@kachkaev Hello, I'd rather keep them separate (versioning and there seem to be a small handful of vault-ui projects out there).

I've pushed a PR for vault-ui to https://github.com/kubernetes/charts/pull/1406

**Review by unguiculus — CHANGES_REQUESTED** [review]
Thanks for the PR. You didn't get standard labels quite right. In your other PR(#1406), they are fine.

**Comment by jpds** [comment]
@unguiculus Ah, this is a `helm create` 2.4 vs 2.5 thing, I've fixed it now. Thanks.

**Comment by jpds** [comment]
@unguiculus Do you know if there's a way to specify a values.yaml just for testing purposes?

The test is failing as there's no default storage backend enabled and I can't enable that as then vault will complain that there are two storage backends when people specify their own.

**Comment by unguiculus** [comment]
A solution for test values is currently being worked on. However, in general, it should be possible to install a chart with default values. Do you think that would be somehow feasible?

**Comment by jpds** [comment]
If a backend isn't provided or if two backends are provided - vault just errors out. Hence why I can't provide a default simple file backend in the YAML, while letting people use their own backend.

**Comment by mumoshu** [comment]
@jpds Not elegant at all but how about introducing the new key `config.json` to accept json like:

```yaml
config:
  json:|
    {
      "listener": {
        "tcp": {
          "address": "0.0.0.0:8200",
          "tls_disable": 1,
      },
      "storage": ...
    }
```

so that we can override the whole config.json?

Also, I guess you can add a conditional dependency to the consul chart via requirements.yaml while making the default `config.json` to assume settings correspond to the consul chart.

File: incubator/vault/values.yaml
Comment (lachie83) [inline incubator/vault/values.yaml]: Can you call out a the URL of where to find a list of supported backends in the config?

**Comment by ipedrazas** [comment]
What I have is a flag that enables the dev mode, which doesn't need backend.

```
          {{- if .Values.vault.mode.dev }}
          command: ["vault", "server", "-dev"]
          {{- else }}
          command: ["vault", "server", "-config", "/conf/config.json"]
          {{- end }}
```

This allows to install the chart and the pod goes into running state just fine.

**Comment by unguiculus** [comment]
Marking as stale. Please update within the next week.

**Comment by jacobkiers** [comment]
@jpds Are you still willing to finish this?

If not, I am interested in seeing this through. I'm new to creating Helm charts though, so I might need a little guidance from a reviewer.

**Comment by unguiculus** [comment]
@jacobkiers Thanks for jumping in. We are happy to assist. You may also ping us on Slack.

**Comment by jacobkiers** [comment]
First  a small question: how should I go about this? Obviously, I cannot push to the branch of @jpds. Should I create a new PR or something?

**Comment by unguiculus** [comment]
@jacobkiers You could actually push to his branch, but it is probably better to fork the PR and create a new one.

@jpds: Last call. Are you willing to continue with this? Otherwise I'll close this PR so @jacobkiers can pick it up.

**Comment by jpds** [comment]
Hello,

I do plan to work on this, I've been very busy recently but should have time to work on this next week.

Thanks,

-------- Message d'origine --------

**Comment by jacobkiers** [comment]
@jpds That's great! I have some small changes already. I'll push my branch
to my own repo so that you don't need to do work that's already done.

Edit: removed mail cruft.

**Comment by jacobkiers** [comment]
@jpds Please see my changes at https://github.com/comandi/charts/commits/vault-initial-chart.

It has two notable changes:
* I implemented the suggestion by @ipedrazas to allow for a dev-server deployment. When that is enabled, all other configuration is ignored.
* The config map containing the Vault config is now tied to the Helm revision. That make sure that when the config is upgraded, de pod will be redeployed with the new configuration. It has the disadvantage that in that case Vault needs to be unsealed again. If there is a better way to do so, feel free to implement that.

I noticed one other thing that I didn't get around to yet: when the liveness and readyness probes are enabled, the pod went into a `CrashLoopBackoff` state. But I don't yet have enough knowledge to find out why.

**Comment by jacobkiers** [comment]
@unguiculus If we want to incorporate some of these changes, then I should probably sign the CLA as well. I'm not sure though whether to do so as an employee or as an individual. I do know that my employer has not yet signed the CLA though.

**Comment by unguiculus** [comment]
@jacobkiers I guess you'll have to find out by yourself how you want to do it.

**Comment by jacobkiers** [comment]
My employer has signed the CLA, so my contributions should be fine to use.

**Comment by tuannvm** [comment]
@jacobkiers @jpds @unguiculus I would love to contribute as well.

@jacobkiers 
> The config map containing the Vault config is now tied to the Helm revision. That make sure that when the config is upgraded, de pod will be redeployed with the new configuration. It has the disadvantage that in that case Vault needs to be unsealed again. If there is a better way to do so, feel free to implement that.

the helm documentation suggests us to use `checksum` to enable roll deployments [here](https://github.com/kubernetes/helm/blob/master/docs/charts_tips_and_tricks.md#automatically-roll-deployments-when-configmaps-or-secrets-change). Furthermore, Helm will not create a lot of configmap with the equivalent revision numbers, that could save us from cleaning up job.

Feel free to take a look on my suggestion [here](https://github.com/tuannvm/charts/commits/tuan-vault-implementation)

**Comment by jpds** [comment]
This should be good to go now.

File: incubator/vault/Chart.yaml
Comment (erickt) [inline incubator/vault/Chart.yaml]: This should be 0.8.1, to match the tagged version.

File: incubator/vault/values.yaml
Comment (erickt) [inline incubator/vault/values.yaml]: Could you document in the README that this setup is insecure by default, and what is the configuration needed to setup TLS on an internal cluster that cannot use Lets Encrypt?

File: incubator/vault/values.yaml
Comment (erickt) [inline incubator/vault/values.yaml]: What do you mean by "rest of this configuration is ignored"? Are you referring to all the options under `vaultmode`? Also, I think it might read a little better if you had:

```yaml
vault:
    dev: true
    config:
        listener:
            ...
        storage:
            ...
```

That gives a more of a namespace to add other vault options that are outside the config file.

File: incubator/vault/README.md
Comment (erickt) [inline incubator/vault/README.md]: Hashicorp recommends running Vault on a [dedicated instance](https://www.vaultproject.io/guides/production.html) to reduce the risk of the machine running Vault getting rooted, and thus allowing an attacker to get to unencrypted secrets. Can you add links to documentation like [kops's instance groups](https://github.com/kubernetes/kops/blob/master/docs/instance_groups.md) and an example config on how to isolate pods to those nodes?

**Review by erickt — CHANGES_REQUESTED** [review]

File: incubator/vault/values.yaml
Comment (ipedrazas) [reply incubator/vault/values.yaml]: Running Vault in dev mode is insecure by design.

Comment (erickt) [reply incubator/vault/values.yaml]: Understood, but given the fact that vault is supposed to store secrets, I think it's worth the effort to keep people from running an insecure vault.

Now that I think about it, we should also add a warning in the status after we install vault if we are in dev mode.

File: incubator/vault/README.md
Comment (jpds) [reply incubator/vault/README.md]: I think running an entire VM instance just for the one container that is running (a rootfs read-only) vault is a little overkill at that point. :-)

File: incubator/vault/values.yaml
Comment (jacobkiers) [inline incubator/vault/values.yaml:42]: Why is dev-mode enabled by default?

Comment (jpds) [reply incubator/vault/values.yaml:42]: So the CI tests can pass with the pod becoming `Ready`.

File: incubator/vault/templates/deployment.yaml
Comment (jacobkiers) [inline incubator/vault/templates/deployment.yaml:29]: I had trouble getting vault up-and-running with the liveness (and/or readyness probe) enabled, because the time between starting vault and unsealing it was too long.

@jpds Did you test this also with a non-dev config? If so, how did you solve that?

File: incubator/vault/values.yaml
Comment (jacobkiers) [reply incubator/vault/values.yaml:42]: Sounds fair.

File: incubator/vault/templates/deployment.yaml
Comment (jpds) [reply incubator/vault/templates/deployment.yaml:29]: Yes, I've used this with before consul and etcd as the backends. Just spun up a new etcd cluster to test it. Didn't have an issue with it starting or unsealing.

Comment (jpds) [reply incubator/vault/templates/deployment.yaml:29]: Stock `incubator/etcd` chart + this as `vault-values.yaml`:

```
vault:
  dev: false
  config:
    storage:
      etcd:
        address: "http://my-release-etcd-0.my-release-etcd.default:2379,my-release-etcd-1.my-release-etcd.default:2379,my-release-etcd-2.my-release-etcd.default:2379"
```

Comment (jacobkiers) [reply incubator/vault/templates/deployment.yaml:29]: Thanks for verifying!

Then it seems there was an error at my end. I'm not that familiar with these probes yet...

**Comment by jacobkiers** [comment]
Seems good to go to me as well.

File: incubator/vault/templates/deployment.yaml
Comment (jpds) [reply incubator/vault/templates/deployment.yaml:29]: I would take a look at the logs from the pod and see if there's a config error.

File: incubator/vault/README.md
Comment (erickt) [reply incubator/vault/README.md]: The risk here is that if Vault is running on the same machine as some other container that's able to get root on the system, it can read the Vault secrets directly out of memory. I've seen some Vault setups that go even more extreme than machine-isolation. For example, Nike created [Cerberus](http://engineering.nike.com/cerberus) that wraps around Vault, and they recommend isolating their setup off in a [dedicated AWS account](http://engineering.nike.com/cerberus/docs/administration-guide/creating-an-environment). I don't think we should hold up this chart for this documentation, but I think it'd be worthwhile adding it in the future.

File: incubator/vault/values.yaml
Comment (erickt) [inline incubator/vault/values.yaml]: This should probably be commented out.

**Review by erickt — COMMENTED** [review]
One minor nit, but beyond that, it looks good to go for me!

I'd love more documentation and warnings about running in dev mode, but we can always add that later.

**Comment by erickt** [comment]
@jpds: Thought of something else while we wait for final review. Could you add anti-affinity to the Deployment, like [this](https://github.com/kubernetes/charts/blob/master/stable/concourse/templates/worker-statefulset.yaml#L79-L88)? That will tell kubernetes to not schedule the Vault replica pods on the same machine.

**Comment by jpds** [comment]
@erickt This anti-affinity definition seems to (strangely) cause a `CrashLoopBackOff`?

**Comment by jpds** [comment]
@unguiculus Apart from the last commit failing - this one should be good to go and it's only your review that's pending.

File: incubator/vault/README.md
Comment (unguiculus) [inline incubator/vault/README.md]: We support the current and the previous release, thus: 1.6+

File: incubator/vault/templates/_helpers.tpl
Comment (unguiculus) [inline incubator/vault/templates/_helpers.tpl]: Use namespaced templates: `vault.name` and `vault.fullname`

See #1785 and https://github.com/kubernetes/helm/pull/2845.

File: incubator/vault/values.yaml
Comment (unguiculus) [inline incubator/vault/values.yaml]: Don't specify defaults for resources.

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: incubator/vault/values.yaml
Comment (kachkaev) [reply incubator/vault/values.yaml]: Hi @unguiculus. Why so? We have a discussion on default `memoryLimit` for traefik in https://github.com/kubernetes/charts/issues/2098, knowing your opinion would be useful.

Comment (unguiculus) [reply incubator/vault/values.yaml]: In general, we recommend to leave this a conscious choice for the user. Users should be able to test a chart out of the box on a system with little resources such as Minikube. If you think you have to specify them (and they are really low anyways in this case), then leave it as it is.

**Review by unguiculus — APPROVED** [review]

**Pull request merged**
